banman/store.go

```diff
@@ -0,0 +1,217 @@
+package banman
+
+import (
+	"encoding/binary"
+	"errors"
+	"net"
+	"time"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+var (
+	// byteOrder is the preferred byte order in which we should write things
+	// to disk.
+	byteOrder = binary.BigEndian
+
+	// banStoreBucket is the top level bucket of the BanStore that will
+	// contain all relevant sub-buckets.
+	banStoreBucket = []byte("ban-store")
+
+	// banBucket is the main index in which we keep track of peers and their
+	// absolute expiration time.
+	//
+	// The key is the peer's host and the value is the absolute expiration
+	// time.
+	banBucket = []byte("ban-index")
+
+	// reasonBucket is an index in which we keep track of why a peer was
+	// banned.
+	//
+	// The key is the peers's host and the value is the Reason.
+	reasonBucket = []byte("reason-index")
+
+	// ErrCorruptedBanStore is an error returned when we attempt to locate
+	// any of the ban-related buckets in the database but are unable to.
+	ErrCorruptedBanStore = errors.New("corrupted ban store")
+)
+
+// BanStatus gathers all of the details regarding a peer's ban status.
+type BanStatus struct {
+	// Banned determines whether the peer is currently banned.
+	Banned bool
+
+	// Reason is the reason for which the peer was banned.
+	Reason Reason
+
+	// Expiration is the absolute time in which the ban will expire.
+	Expiration time.Time
+}
+
+// BanStore is the store responsible for maintaining records of banned peers.
+type BanStore interface {
+	// BanPeer creates a ban record for the peer within the BanStore for the
+	// given duration. A reason can also be provided to note why the peer is
+	// being banned. The record will exist until a call to BanStatus is made
+	// after the ban expiration.
+	BanPeer(string, Reason, time.Duration) error
```

Comment:
I think we should ban based on `net.IP`, rather than a `string`. In the future, along with eclipse attack prevention measures, we may start to ban on a netblocks (IP address ranges). Example type we may grow to utilize: https://golang.org/pkg/net/#IPNet. If there's no `Mask` field set there, then it just target a single IP address.

Reply:
Doesn't seem like there's a difference between using `net.IP` and `string`? Though if we know there's a potential use case for `net.IPNet`, then I can update the PR to use that now instead.

Reply:
I outlined a precise use case above. Beyond that it adds stronger typing making the usage patterns of the new store more explicit. We may also want to have distinct bannig heuristics for say IPv4 vs IPv6 vs Tor. IPv4 is the most scarce so net block banning has a larger impact. Conversely for Tor hidden service addresses, making a new one is mostly costless (though rate limited a bit).

Reply:
Pushed a new version that relies on `net.IPNet` as discussed, though it doesn't currently support Tor addresses due to the standard library only supporting IPv4 and IPv6 addresses. This can be done, but will require some more logic which could be done in a follow-up.